src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/AggregateFunction.java

```diff
@@ -0,0 +1,60 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types;
+
+import static com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.BaseType.ES_TYPE;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.BaseType.INTEGER;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.BaseType.NUMBER;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.Generic.T;
+
+/**
+ * Aggregate function
+ */
+public enum AggregateFunction implements TypeExpression {
+    COUNT(
+        func().to(INTEGER), // COUNT(*)
+        func(ES_TYPE).to(INTEGER)
+    ),
+    MAX(func(T(NUMBER)).to(T)),
```

Comment:
I think the purpose of func(T(NUMBER)).to(T) is to bind the return type to the actual paramater type. could we doing this by func(NUMBER).to(ArgType(0))?

Reply:
Yes, let me come up with a better form for generic type.

Reply:
Agree it's not the best way. But keep it as it is and add comments for now. I'm planning to refactor it in next phase along with documentation.